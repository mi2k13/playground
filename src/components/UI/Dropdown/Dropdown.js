import React from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { List } from 'immutable';
// components
import DropdownItem from '../DropdownItem/DropdownItem';
import Icon from '../Icon/Icon';
// styles
import styles from './Dropdown.css';
import classnames from 'classnames/bind';
const cx = classnames.bind(styles);

// ============================================

class Dropdown extends React.Component {
  constructor() {
    super();

    this.state = {
      open: false,
    };

    this.toggleMenu = this.toggleMenu.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('click', this.handleClickOutside, true);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClickOutside, true);
  }

  handleClickOutside(event) {
    const domNode = ReactDOM.findDOMNode(this);

    if ((!domNode || !domNode.contains(event.target))) {
        this.setState({ open : false });
    }
  }

  toggleMenu() {
    this.setState({ open: !this.state.open });
  }

  render() {
    const {
      currentLabel,
      handleChange,
      options,
      style,
    } = this.props;

    const divStyle = cx(
      styles.root,
      style,
      {
        open: this.state.open,
      }
    );

    return (
      <span
        className={divStyle}
        onClick={this.toggleMenu}
      >
        {/* LABEL */}
        <span className={styles.labelContainer}>
          <span className={styles.label}>
            {currentLabel}
          </span>
          <Icon
            symbol="arrow"
            style={styles.icon}
          />
        </span>

        {/* MENU */}
        <div className={styles.menu}>
          <ul>
            {options.map((option, id) => (
                <DropdownItem
                  key={id}
                  isActive={option.get('label') === currentLabel}
                  item={option}
                  handleClick={handleChange}
                />
              )
            )}
          </ul>
        </div>
      </span>
    )
  }
}

Dropdown.propTypes = {
  currentLabel: PropTypes.string,
  handleChange: PropTypes.func,
  options: PropTypes.instanceOf(List).isRequired,
  style: PropTypes.string,
};

export default Dropdown;
