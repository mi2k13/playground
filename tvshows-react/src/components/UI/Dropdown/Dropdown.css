.root {
  position: relative;
  cursor: pointer;
}
.root.open .menu {
  display: block;
}

.labelContainer {
  padding: .4em .8em;
  border-radius: $baseBorderRadius;
  background: $grayLight3;
}
.label {
  vertical-align: baseline;
}

.icon {
  margin-left: $gapSmall;
  transform: rotate(90deg);
}

.menu {
  position: absolute;
  top: 100%;
  left: 0;
  display: none;
  margin-top: .4em;
  border-radius: $baseBorderRadius;
  background: $grayLight2;
}
