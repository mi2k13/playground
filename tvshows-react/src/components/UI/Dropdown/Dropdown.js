import React from 'react';
import { List } from 'immutable';
// components
import DropdownItem from '../DropdownItem/DropdownItem';
import Icon from '../Icon/Icon';
// styles
import styles from './Dropdown.css';
import classnames from 'classnames/bind';
const cx = classnames.bind(styles);

// ============================================

class Dropdown extends React.Component {
  constructor() {
    super();

    this.state = {
      open: false,
    };

    this.toggleMenu = this.toggleMenu.bind(this);
  }

  toggleMenu() {
    this.setState({ open: !this.state.open });
  }

  render() {
    const {
      currentLabel,
      handleChange,
      options,
      style,
    } = this.props;

    const divStyle = cx(
      styles.root,
      style,
      {
        open: this.state.open,
      }
    );

    return (
      <div
        className={divStyle}
        onClick={this.toggleMenu}
      >
        {/* LABEL */}
        <span className={styles.labelContainer}>
          <span className={styles.label}>
            {currentLabel}
          </span>
          <Icon
            symbol="arrow"
            style={styles.icon}
          />
        </span>

        {/* MENU */}
        <ul className={styles.menu}>
          {options.map((option, id) => (
              <DropdownItem
                key={id}
                isActive={option.get('label') === currentLabel}
                item={option}
                handleClick={handleChange}
              />
            )
          )}
        </ul>
      </div>
    )
  }
}

Dropdown.propTypes = {
  currentLabel: React.PropTypes.string,
  handleChange: React.PropTypes.func,
  options: React.PropTypes.instanceOf(List).isRequired,
  style: React.PropTypes.string,
};

export default Dropdown;
