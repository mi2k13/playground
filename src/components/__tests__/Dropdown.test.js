import React from 'react';
import renderer from 'react-test-renderer';
import { List, Map } from 'immutable';
import Dropdown from '../UI/Dropdown/Dropdown';

// ============================================

test('Renders Dropdown menu', () => {
  const tree = renderer.create(
    <Dropdown options={List([
      Map({
        value: 1,
        label: 'First item'
      }),
      Map({
        value: 2,
        label: 'Second item'
      }),
    ])} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Toggle Dropdown menu', () => {
  const tree = renderer.create(
    <Dropdown options={List([
      Map({
        value: 1,
        label: 'First item'
      }),
      Map({
        value: 2,
        label: 'Second item'
      }),
    ])} />
  ).toJSON();
  tree.props.onClick();
  expect(tree).toMatchSnapshot();
});
