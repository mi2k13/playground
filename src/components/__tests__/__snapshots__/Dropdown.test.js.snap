exports[`test Renders Dropdown menu 1`] = `
<span
  className=""
  onClick={[Function]}>
  <span
    className={undefined}>
    <span
      className={undefined} />
    <svg
      className="regular">
      <use
        xlinkHref="#icon-arrow" />
    </svg>
  </span>
  <div
    className={undefined}>
    <ul>
      <li
        className=""
        onClick={[Function]}>
        First item
      </li>
      <li
        className=""
        onClick={[Function]}>
        Second item
      </li>
    </ul>
  </div>
</span>
`;

exports[`test Toggle Dropdown menu 1`] = `
<span
  className=""
  onClick={[Function]}>
  <span
    className={undefined}>
    <span
      className={undefined} />
    <svg
      className="regular">
      <use
        xlinkHref="#icon-arrow" />
    </svg>
  </span>
  <div
    className={undefined}>
    <ul>
      <li
        className=""
        onClick={[Function]}>
        First item
      </li>
      <li
        className=""
        onClick={[Function]}>
        Second item
      </li>
    </ul>
  </div>
</span>
`;
